File: Classes/Model classes/CountryCode.m
/*
	MyFlightbook for iOS - provides native access to MyFlightbook
	pilot's logbook
 Copyright (C) 2017 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  CountryCode.m
//  MFBSample
//
//  Created by Eric Berman on 11/7/14.
//
//

#import "CountryCode.h"
#import "MFBAppDelegate.h"
#import "Util.h"
#import <sqlite3.h>

@implementation CountryCode

@synthesize LocaleCode, CountryName, Prefix, ID;

static NSMutableArray * rgAllCountryCodes = nil;

- (instancetype) initFromRow:(sqlite3_stmt *) row
{
    if (self = [super init])
    {
        self.ID = sqlite3_column_int(row, 0);
        self.Prefix = [NSString stringFromCharsThatCouldBeNull:(char *) sqlite3_column_text(row, 1)];
        self.LocaleCode = [NSString stringFromCharsThatCouldBeNull:(char *) sqlite3_column_text(row, 2)];
        self.CountryName = [NSString stringFromCharsThatCouldBeNull:(char *) sqlite3_column_text(row, 3)];
    }
    return self;
}

- (NSString *) description
{
    return [NSString stringWithFormat:@"%ld: %@ %@ %@", (long) self.ID, self.Prefix, self.LocaleCode, self.CountryName];
}

+ (NSArray *) AllCountryCodes
{
    if (rgAllCountryCodes != nil)
        return rgAllCountryCodes;

    rgAllCountryCodes = [[NSMutableArray alloc] init];
    sqlite3 * db = mfbApp().getdb;
    sqlite3_stmt * sqlCountryCodes = nil;
    if (sqlite3_prepare(db, [@"SELECT * FROM countrycodes " cStringUsingEncoding:NSASCIIStringEncoding], -1, &sqlCountryCodes, NULL) != SQLITE_OK)
        NSLog(@"Error: failed to prepare statement with message '%s'.", sqlite3_errmsg(db));
    
    while (sqlite3_step(sqlCountryCodes) == SQLITE_ROW)
        [rgAllCountryCodes addObject:[[CountryCode alloc] initFromRow:sqlCountryCodes]];
    
    sqlite3_finalize(sqlCountryCodes);
    return rgAllCountryCodes;
}

+ (CountryCode *) BestGuessForLocale:(NSString *) locale
{
    NSArray * rg = [CountryCode AllCountryCodes];

    if (locale != nil && locale.length > 1)
    {
        for (CountryCode * cc in rg)
            if ([cc.LocaleCode compare:locale options:NSCaseInsensitiveSearch] == NSOrderedSame)
                return cc;
    }
    
    return rg[0];
}

+ (CountryCode *) BestGuessForCurrentLocale
{
    return [CountryCode BestGuessForLocale:[[NSLocale currentLocale] objectForKey:NSLocaleCountryCode]];
}

// return the longest country code that is a prefix for the given tail number
+ (CountryCode *) BestGuessPrefixForTail:(NSString *) szTail
{
    CountryCode * result = nil;
    NSInteger maxLength = 0;
    
    NSArray * rg = [CountryCode AllCountryCodes];
    for (NSInteger i = [rg count] - 1; i >= 0; i--)
    {
        CountryCode * cc = ((CountryCode *) rg[i]);
        NSString * szPref = cc.Prefix;
        if ([szTail hasPrefix:szPref] && [szPref length] > maxLength)
        {
            result = cc;
            maxLength = [szPref length];
        }
    }
    
    return result;
}

@end

File: Classes/Model classes/Logbook Model/CountryCode.swift
//
//  CountryCode.swift
//  MyFlightbook
//
//  Created by Eric Berman on 3/9/23.
//

import Foundation
import SQLite3

@objc public class CountryCode : NSObject {
    private let ID : Int
    private let LocaleCode : String
    @objc public let Prefix : String
    private let CountryName : String
    
    private static var rgAllCountryCodes : [CountryCode] = []
    
    private init(_ row: OpaquePointer!) {
        ID = Int(sqlite3_column_int(row, 0))
        var sz = sqlite3_column_text(row, 1)
        Prefix = sz == nil ? "" : String(cString: sz!)
        sz = sqlite3_column_text(row, 2)
        LocaleCode = sz == nil ? "" : String(cString: sz!)
        sz = sqlite3_column_text(row, 3)
        CountryName = sz == nil ? "" : String(cString: sz!)
        super.init()
    }
    
    public override var description : String {
        get {
            return String(format: "%d: %@ %@ %@", ID, Prefix, LocaleCode, CountryName)
        }
    }
    
    private static var AllCountryCodes : [CountryCode] {
        get {
            if (!rgAllCountryCodes.isEmpty) {
                return rgAllCountryCodes
            }
            
            var sqlCountryCodes : OpaquePointer?
            let db = MFBSqlLite.current
            
            if (sqlite3_prepare(db, "SELECT * FROM countrycodes ORDER BY Prefix ASC".cString(using: .ascii), -1, &sqlCountryCodes, nil) != SQLITE_OK) {
                NSLog("Error: failed to prepare statement with message '%s'.", sqlite3_errmsg(db))
            }
            
            while (sqlite3_step(sqlCountryCodes) == SQLITE_ROW) {
                rgAllCountryCodes.append(CountryCode(sqlCountryCodes))
            }
            
            sqlite3_finalize(sqlCountryCodes)
            return rgAllCountryCodes
        }
    }
    
    @objc public static func BestGuessForLocale(_ locale : String) -> CountryCode {
        let rg = AllCountryCodes
        
        let rgMatches = rg.filter() { $0.LocaleCode.compare(locale, options:.caseInsensitive) == .orderedSame }

        if (rgMatches.isEmpty) {
            return rg[0]
        }
        
        let rgSorted = rgMatches.sorted { cc1, cc2 in
            return cc1.Prefix.compare(cc2.Prefix, options: .caseInsensitive) != .orderedDescending
        }
        
        return rgSorted[0]
    }
    
    @objc public static func BestGuessForCurrentLocale() -> CountryCode {
        return BestGuessForLocale((Locale.current as NSLocale).object(forKey: .countryCode) as! String)
    }
    
    // return the longest country code that is a prefix for the given tail number
    @objc public static func BestGuessPrefixForTail(_ szTail : String) -> CountryCode? {
        var result : CountryCode? = nil
        var maxLength = 0
        
        let rg = AllCountryCodes
        var idx = rg.count - 1
        for cc in rg {
            if (szTail.hasPrefix(cc.Prefix) && cc.Prefix.count > maxLength) {
                result = cc
                maxLength = cc.Prefix.count
            }
            idx -= 1
        }
        
        return result
    }
}
